clang/lib/Driver/ToolChains/BareMetal.cpp

```diff
@@ -382,38 +382,6 @@ void BareMetal::AddClangCXXStdlibIncludeArgs(const ArgList &DriverArgs,
   }
 }
 
-void BareMetal::AddCXXStdlibLibArgs(const ArgList &Args,
-                                    ArgStringList &CmdArgs) const {
-  switch (GetCXXStdlibType(Args)) {
-  case ToolChain::CST_Libcxx:
-    CmdArgs.push_back("-lc++");
-    if (Args.hasArg(options::OPT_fexperimental_library))
-      CmdArgs.push_back("-lc++experimental");
-    CmdArgs.push_back("-lc++abi");
```

Comment:
It looks like the generic version of this doesn't include `-lc++abi`, while this one does. Is that an important difference? It seems like most of the derived toolchains include this in their overrides.

Reply:
This could be a static/dynamic linking thing. If libc++ is a shared library it would have a DT_NEEDED libc++abi (or the GCC equivalent).

For static links I think we'd need the explicit `-lc++abi` or the GNU equivalent.

If I can find some time I'll see if we can try this patch out on our exceptions/rtti library variant as that would need objects from libc++abi.

Reply:
Yes I can confirm that leaving out -lc++abi does cause build failures with our C++ samples with undefined symbols from libc++abi.

The lld command (paths shortened for brevity) coming out of `clang++ --target=armv6m-none-eabi -march=armv6m -mfpu=none -mfloat-abi=soft -lcrt0-semihost -lsemihost -g -T ../../ldscripts/microbit.ld -o hello-exn.elf hello-exn.cpp`
```
"bin/ld.lld" -lcrt0-semihost -lsemihost /tmp/hello-exn-5244ae.o -Bstatic -EL -T ../../ldscripts/microbit.ld -Lbin/../lib/clang-runtimes/arm-none-eabi/armv6m_soft_nofp_exn_rtti/lib -Llib/clang/20/lib/armv6m-unknown-none-eabi -Lbin/../lib/clang-runtimes/arm-none-eabi/armv6m_soft_nofp_exn_rtti/lib -lc++ -lm b\
in/../lib/clang-runtimes/arm-none-eabi/armv6m_soft_nofp_exn_rtti/lib/libclang_rt.builtins.a -lc --target2=rel -o hello-exn.elf
ld.lld: error: undefined symbol: operator new(unsigned int)
...
ld.lld: error: undefined symbol: __cxa_allocate_exception
...
```

Reply:
The problem with the current implementation is that it assumes that libc++abi is used as the C++ ABI library, but libc++ supports [a number of different C++ ABI library implementations](https://github.com/llvm/llvm-project/blob/c7c5e05389292da7e5a87e3d1d3ef08021911a53/libcxx/CMakeLists.txt#L234), including `none` which may be desirable for some of the baremetal use cases and we use it in some of our embedded targets.

There are two solutions for pulling in the right C++ ABI library without needing to explicitly pass `-lc++abi` to the linker that are already used by other platforms:

1. Replace `libc++.so` and `libc++.a` with an input linker script that pulls in the right C++ ABI library; libc++ build can [generate one for `libc++.so`](https://github.com/llvm/llvm-project/blob/c7c5e05389292da7e5a87e3d1d3ef08021911a53/libcxx/CMakeLists.txt#L266), this could be extended to support `libc++.a` if needed.
2. You can merge `libc++abi.a` into `libc++.a` so you don't need a separate `-lc++abi`, [this is already supported by libc++ build](https://github.com/llvm/llvm-project/blob/c7c5e05389292da7e5a87e3d1d3ef08021911a53/libcxx/CMakeLists.txt#L244).

The advantage of either of these solutions is that Clang driver can remain agnostic to which C++ ABI library the vendor built libc++ against.

@smithp35 Would it be feasible to migrate your toolchain to use either of these solutions?

Reply:
libc++abi is intended to be an implementation detail of libc++. I don't have authority over the Clang Driver, but it makes sense to me that the Driver would only link against libc++ and assume that this provides the necessary symbols. If that means libc++ needs to merge libc++abi into itself or something like that, that's IMO the right thing to do.

Reply:
> 
> @smithp35 Would it be feasible to migrate your toolchain to use either of these solutions?

It would definitely be feasible. I'm hoping that we can make a solution drop out of the existing runtimes build perhaps with an additional CMake flag. For example if we provide an extra option to the runtimes build then it either spits out a libc++ linker script fragment (my preference) or merge libc++ and libc++abi libraries (this should work as long as libc++abi's objects are inserted after libc++).

Reply:
> spits out a libc++ linker script fragment (my preference)

We already have this and it's used on platforms like Darwin, Fuchsia or Linux:

https://github.com/llvm/llvm-project/blob/3d08ade7bd32f0296e0ca3a13640cc95fa89229a/libcxx/CMakeLists.txt#L269-L271

> merge libc++ and libc++abi libraries (this should work as long as libc++abi's objects are inserted after libc++)

We already have this and it's what we've been using in our baremetal build for example:

https://github.com/llvm/llvm-project/blob/3d08ade7bd32f0296e0ca3a13640cc95fa89229a/libcxx/CMakeLists.txt#L247-L253

Reply:
Thanks for pointing those out, I was under the impression that they had to be developed first. Although I'm not the Driver owner either I can try those out and get back to you tomorrow (apologies its a bit late in the day) with a second approval.